File: rp_engine/models/custom_state.py
"""Custom state tracking models."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class CustomStateSchema(BaseModel):
    id: str
    rp_folder: str | None = None
    category: str
    name: str
    data_type: str  # "number", "text", "list", "object"
    config: dict | None = None
    belongs_to: str | None = None  # NULL = scene-level, "character" = per-character
    inject_as: str = "hidden"  # "stat_block", "inventory_list", "note", "hidden"
    display_order: int = 0


class CustomStateSchemaCreate(BaseModel):
    id: str
    rp_folder: str
    category: str
    name: str
    data_type: str
    config: dict | None = None
    belongs_to: str | None = None
    inject_as: str = "hidden"
    display_order: int = 0


class CustomStateValue(BaseModel):
    schema_id: str
    entity_id: str | None = None
    value: str | dict | list | int | float | None = None
    exchange_number: int | None = None
    changed_by: str | None = None
    reason: str | None = None


class CustomStateSet(BaseModel):
    schema_id: str
    rp_folder: str
    branch: str = "main"
    entity_id: str | None = None
    value: str | dict | list | int | float
    reason: str | None = None


class PresetInfo(BaseModel):
    name: str
    description: str | None = None
    schema_count: int = 0


class CustomStateSnapshot(BaseModel):
    """All current custom state values for an RP."""
    schemas: list[CustomStateSchema] = []
    values: list[CustomStateValue] = []


class PCCustomStateItem(BaseModel):
    schema_id: str
    category: str
    name: str
    data_type: str
    value: Any = None
    display_format: str  # inject_as


class PCStateResponse(BaseModel):
    character: str
    location: str | None = None
    emotional_state: str | None = None
    conditions: list[str] = []
    custom_state: list[PCCustomStateItem] = []

File: frontend/src/lib/types/customState.ts
export interface CustomStateSchema {
	id: string;
	rp_folder: string | null;
	category: string;
	name: string;
	data_type: string;
	config: Record<string, unknown> | null;
	belongs_to: string | null;
	inject_as: string;
	display_order: number;
}

export interface CustomStateValue {
	schema_id: string;
	entity_id: string | null;
	value: unknown;
	exchange_number: number | null;
	changed_by: string | null;
	reason: string | null;
}

export interface CustomStateSnapshot {
	schemas: CustomStateSchema[];
	values: CustomStateValue[];
}

export interface PresetInfo {
	name: string;
	description: string | null;
	schema_count: number;
}
